### ISSUE TITLE
C++ backend should be stricter with UTF-8

### ISSUE THREAD

**Issue by sffc** [rust-diplomat/diplomat#471]
https://github.com/unicode-org/icu4x/pull/4837 landed code that reads:

```
   * Warning: Passing ill-formed UTF-8 is undefined behavior (and may be memory-unsafe).
   */
  ICU4XBidiInfo for_text(const std::string_view text, uint8_t default_level) const;
```

But there's nothing enforcing that the `std::string_view` is UTF-8 other than this comment. I think that API is unsafe. Also, there's no similar comment in C, which must have the same issue.

Better would be to do something like:

1. Enforce UTF-8 on the C++ side by having a `diplomat::utf8_string_view` that has both a safe and unsafe constructor
2. Use C++20 `u8string_view`

**Comment by robertbastian** [comment]
That API is unsafe in the Rust sense, but this is pretty standard for C++. C++ doesn't track UTF-8 well-formedness at the type level like Rust does, both `string_view` and `u8string_view` can be created with invalid UTF-8. Introducing `diplomat::utf8_string_view` adds an additional iteration of the string, which we want to avoid. I think the solution for ICU4X is to rewrite this method in a way that accepts invalid UTF-8.

**Comment by Manishearth** [comment]
Yeah, we explicitly give the bridge code control over this because there is no strong norm in C++. I don't consider this to be a Diplomat bug, this is potentially an ICU4X bug.

We should have this autogenerate docs about ill formedness.

**Comment by robertbastian** [comment]
Regarding C, there are currently no docs in C or C2 to add this to. I definitely want this in C docs once we have them.

**Comment by sffc** [comment]
> I think the solution for ICU4X is to rewrite this method in a way that accepts invalid UTF-8.

Why did we merge the PR that removed the assertion and unlocked the unsafe behavior, then?

> u8string_view can be created with invalid UTF-8

How? I think you need to `reinterpret_cast` to do this, which is the closest C++ gets to an `unsafe` block.

> Introducing diplomat::utf8_string_view adds an additional iteration of the string, which we want to avoid.

It can have an cheap constructor from `std::u8string_view`, and if people have a `char*` that they are confident is UTF-8, they can `reinterpret_cast` to `char8_t*` and then use the cheap constructor.

---

My overall feeling is that it is Diplomat's duty to make sure that if it gives a `&str` to Rust code, that it has done its due diligence to make sure that it is valid UTF-8. My position is that a comment in a doc string is _not_ due diligence.

**Comment by sffc** [comment]
Aside from whether or not `char8_t*` has the invariants we want it to have:

My position is that a doc string is not a sufficient safeguard to protect against UB, even in C++.

That is a position one can disagree with. However, it's long been the practice in ICU4C and elsewhere to, at the very least, not take for granted the well-formedness of input UTF-8 strings. This is why ICU4C has had both safe and `UNSAFE` UTF-8 iterators in [utf8.h](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/utf8_8h.html), long before Rust existed. These UTF-8 iteration macros are the gold standard for C++ programmers. If you are 100% certain that you are dealing with UTF-8, you use the `UNSAFE` macros; otherwise, you use the default (safe) ones.

I would be okay with any of the following:

1. Introduce `diplomat::utf8_string_view` that has both safe and `unsafe` constructors. The interaction with `char8_t*`, if any, can be discussed later.
2. Make the C++ backend refuse to generate bindings for functions taking a `&str` argument.
3. Make the C++ backend affix a suffix such as `_utf8_unsafe` to functions taking a `&str` argument.

My personal preference, and the least amount of work, for the time being would be (2), and all ICU4X functions should take `DiplomatStr` and internally perform the required checks/conversions.

**Comment by Manishearth** [comment]
I think (2) should be an option during generation (it's fine to have the default you suggest). I fundamentally think this is a choice that users of Diplomat should make.

Not really a fan of (1). I'm okay with a stdlib-based solution like u8string_view even if it isn't perfect, but it would have to be opt-in because C++20.

**Comment by sffc** [comment]
I'm fine with this being an option during generation. We should make it the default and/or use it in ICU4X.

**Comment by robertbastian** [comment]
> Why did we merge the PR that removed the assertion and unlocked the unsafe behavior, then?

The PR changes a panic into an unsafety. Our value proposition is that we're panic-free, and C++ is used to unsafety.

**Comment by sffc** [comment]
Discussed on 2024-04-30:

- No agreement on the general Diplomat behavior. @sffc asked others to provide ecosystem examples of documentation-based unsafe UTF-8 (with potential reads past end of buffer).
- For ICU4X specifically, we will continue migrating our FFI to accept potentially ill-formed UTF-8. Two of three functions use ICU4X itself and are easy to migrate. The third interacts with Bi-Di.
- For the Bi-Di function, further discussion is required regarding handling of both potentially ill-formed UTF-8 and potentially ill-formed UTF-16. @Manishearth and @robertbastian to discuss.
    - One idea: we will have version `for_text` (which takes `&DiplomatStr` and stops processing after reaching invalid UTF-8 [@robertbastian not fond of this]) and `for_text_valid_utf8` (which takes `&str`, disabled in C++, and renamed in other targets). We might migrate the `for_text` potentially ill-formed behavior into `unicode_bidi` at some point.

LGTM: @Manishearth @sffc @robertbastian @younies

**Comment by robertbastian** [comment]
I think this was fixed by #482, discussion in #475
